fenix/app/src/main/java/org/mozilla/fenix/telemetry/TelemetryMiddleware.kt

```diff
@@ -20,27 +20,37 @@ import mozilla.components.concept.base.crash.CrashReporting
 import mozilla.components.lib.state.Middleware
 import mozilla.components.lib.state.MiddlewareContext
 import mozilla.components.support.base.log.logger.Logger
+import mozilla.telemetry.glean.internal.TimerId
 import mozilla.telemetry.glean.private.NoExtras
 import org.mozilla.fenix.Config
 import org.mozilla.fenix.GleanMetrics.Events
 import org.mozilla.fenix.GleanMetrics.Metrics
 import org.mozilla.fenix.components.metrics.Event
 import org.mozilla.fenix.components.metrics.MetricController
 import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.nimbus.FxNimbus
 import org.mozilla.fenix.utils.Settings
 import org.mozilla.fenix.GleanMetrics.EngineTab as EngineMetrics
 
+private const val PROGRESS_COMPLETE = 100
+
 /**
  * [Middleware] to record telemetry in response to [BrowserAction]s.
  *
  * @property settings reference to the application [Settings].
  * @property metrics [MetricController] to pass events that have been mapped from actions
+ * @property nimbusSearchEngine The Nimbus search engine.
+ * @property searchState Map that stores the [TabSessionState.id] & [TimerId].
+ * @property timerIdFunc Fun to provide the Metric [TimerId].
  */
 class TelemetryMiddleware(
     private val context: Context,
     private val settings: Settings,
     private val metrics: MetricController,
     private val crashReporting: CrashReporting? = null,
+    private val nimbusSearchEngine: String = FxNimbus.features.searchExtraParams.value().searchEngine,
+    private val searchState: MutableMap<String, TimerId> = mutableMapOf(),
+    private val timerIdFunc: () -> TimerId = Metrics.searchPageLoadTime::start,
```

Comment:
nit: Since we already know that this is a function, maybe instead of `timerIdFunc` we could use something like `createTimer` instead? Up to you :)